// @sipheron/vdr-core/dist/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SipHeron = void 0;
const config_1 = require("./config");
const http_1 = require("./http");
const hash_1 = require("../hash");
const errors_1 = require("../errors");
/**
 * Main SipHeron VDR client.
 *
 * @example
 * const sipheron = new SipHeron({ apiKey: 'your-key', network: 'devnet' })
 * const anchor = await sipheron.anchor({ file, name: 'Contract v1' })
 * const result = await sipheron.verify({ file })
 */
class SipHeron {
    constructor(config) {
        this.config = (0, config_1.resolveConfig)(config);
        this.http = (0, http_1.createHttpClient)(this.config);
    }
    /**
     * Anchor a document's SHA-256 fingerprint permanently to Solana.
     *
     * When file is provided, hashing happens locally before anything
     * is transmitted — the document bytes never leave your machine.
     *
     * @param options - Anchor options
     * @returns AnchorResult with transaction details and verification URL
     *
     * @throws {ValidationError} If neither file nor hash is provided
     * @throws {AuthenticationError} If API key is invalid
     * @throws {RateLimitError} If per-second rate limit is exceeded
     * @throws {QuotaExceededError} If monthly quota is exceeded
     * @throws {NetworkError} If network request fails
     *
     * @example
     * const anchor = await sipheron.anchor({
     *   file: fs.readFileSync('./contract.pdf'),
     *   name: 'Service Agreement — Acme Corp',
     * })
     * console.log(anchor.verificationUrl)
     */
    async anchor(options) {
        // Validate input
        if (!options.file && !options.hash) {
            throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string). Neither was provided.');
        }
        if (options.file && options.hash) {
            throw new errors_1.ValidationError('Provide either file or hash — not both.');
        }
        // Compute hash locally if file provided
        let hash;
        if (options.file) {
            hash = await (0, hash_1.hashDocument)(options.file);
        }
        else {
            hash = (0, hash_1.normalizeHash)(options.hash);
            if (!(0, hash_1.isValidHash)(hash)) {
                throw new errors_1.ValidationError('Invalid hash format. Must be a 64-character lowercase hex string (SHA-256).');
            }
        }
        // Build request headers
        const headers = {};
        if (options.idempotencyKey) {
            headers['Idempotency-Key'] = options.idempotencyKey;
        }
        if (!this.config.apiKey && this.config.network !== 'devnet') {
            throw new errors_1.AuthenticationError('apiKey is required to anchor on mainnet.');
        }
        const endpoint = !this.config.apiKey ? '/api/playground/anchor' : '/api/hashes';
        // Post to API
        const response = await (0, http_1.withRetry)(() => this.http.post(endpoint, {
            hash,
            filename: options.name || options.metadata?.name || null,
            metadata: options.name || options.metadata?.name || null,
            ...(options.metadata && { tags: Object.keys(options.metadata) }),
        }, { headers }), this.config.retries);
        return this._mapAnchorResponse(response.data);
    }
    /**
     * Anchor multiple documents in a batch.
     * Processes up to 500 documents per call.
     *
     * @param options - Batch anchor options
     * @returns BatchAnchorResult with individual results
     */
    async anchorBatch(options) {
        if (!this.config.apiKey) {
            throw new errors_1.AuthenticationError('apiKey is required for batch anchoring. Use single anchor() for devnet playground.');
        }
        if (!options.documents || options.documents.length === 0) {
            throw new errors_1.ValidationError('documents array cannot be empty');
        }
        if (options.documents.length > 500) {
            throw new errors_1.ValidationError('Maximum 500 documents per batch');
        }
        const results = [];
        let successful = 0;
        let failed = 0;
        for (const doc of options.documents) {
            try {
                const result = await this.anchor(doc);
                results.push(result);
                successful++;
            }
            catch (error) {
                if (options.stopOnError)
                    throw error;
                results.push({
                    error: error.message,
                    input: doc,
                });
                failed++;
            }
        }
        return { results, successful, failed, total: options.documents.length };
    }
    /**
     * Verify a document against its blockchain anchor.
     *
     * When file is provided, hashing happens locally — the document
     * bytes are never transmitted.
     *
     * @param options - Verify options
     * @returns VerificationResult with authentic boolean and anchor details
     *
     * @throws {ValidationError} If neither file nor hash is provided
     * @throws {AnchorNotFoundError} If no anchor record exists for this hash
     * @throws {AnchorRevokedError} If the anchor has been revoked
     *
     * @example
     * const result = await sipheron.verify({ file })
     * if (result.authentic) {
     *   console.log('Document is authentic')
     * } else {
     *   console.log('Document has been modified')
     * }
     */
    async verify(options) {
        if (!options.file && !options.hash) {
            throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string). Neither was provided.');
        }
        // Compute hash locally if file provided
        let hash;
        if (options.file) {
            hash = await (0, hash_1.hashDocument)(options.file);
        }
        else {
            hash = (0, hash_1.normalizeHash)(options.hash);
            if (!(0, hash_1.isValidHash)(hash)) {
                throw new errors_1.ValidationError('Invalid hash format. Must be a 64-character hex string (SHA-256).');
            }
        }
        const endpoint = !this.config.apiKey ? '/api/playground/verify' : '/api/verify';
        const response = await (0, http_1.withRetry)(() => this.http.post(endpoint, { hash }), this.config.retries);
        const data = response.data;
        // Handle revoked
        if (data.status === 'REVOKED') {
            throw new errors_1.AnchorRevokedError(hash, data.anchor?.revokedAt, data.anchor?.revocationNote);
        }
        return {
            authentic: data.authentic === true,
            status: this._mapVerifyStatus(data.status),
            hash,
            verifiedAt: data.verified_at,
            anchor: data.anchor ? this._mapAnchorResponse(data) : undefined,
            anchoredHash: data.anchor?.hash,
        };
    }
    /**
     * Get the current status of an anchor by hash.
     *
     * @param hash - 64-character SHA-256 hash
     * @returns AnchorResult with current status
     */
    async getStatus(hash) {
        const normalized = (0, hash_1.normalizeHash)(hash);
        if (!(0, hash_1.isValidHash)(normalized)) {
            throw new errors_1.ValidationError('Invalid hash format.');
        }
        const endpoint = !this.config.apiKey
            ? `/api/hashes/public/${normalized}`
            : `/api/hashes/${normalized}/status`;
        const response = await (0, http_1.withRetry)(() => this.http.get(endpoint), this.config.retries);
        return this._mapAnchorResponse(response.data);
    }
    /**
     * List anchors for the authenticated organization.
     *
     * @param options - Filtering and pagination options
     */
    async list(options) {
        if (!this.config.apiKey) {
            throw new errors_1.AuthenticationError('apiKey is required to list hashes.');
        }
        const params = new URLSearchParams();
        if (options?.page)
            params.set('page', String(options.page));
        if (options?.limit)
            params.set('limit', String(options.limit));
        if (options?.status)
            params.set('status', options.status);
        if (options?.search)
            params.set('search', options.search);
        const response = await (0, http_1.withRetry)(() => this.http.get(`/api/hashes?${params}`), this.config.retries);
        const data = response.data;
        const records = (data.records || data.hashes || []).map((r) => this._mapAnchorResponse({ anchor: r }));
        return {
            records,
            total: data.total || 0,
            page: data.page || 1,
            pages: data.pages || 1,
        };
    }
    /**
     * Verify a pre-computed hash without providing the document.
     * Use this when you already have the SHA-256 hash.
     *
     * @param hash - 64-character SHA-256 hash
     * @returns VerificationResult
     */
    async verifyHash(hash) {
        return this.verify({ hash });
    }
    // ── Private helpers ──
    _mapAnchorResponse(data) {
        const anchor = (data.anchor || data.record || data);
        const blockchain = (data.blockchain || {});
        const hash = String(anchor.hash || '');
        const txSig = String(anchor.txSignature || blockchain.txSignature || '');
        return {
            id: String(anchor.id || ''),
            hash,
            transactionSignature: txSig,
            blockNumber: Number(anchor.blockNumber || blockchain.blockNumber || 0),
            timestamp: String(anchor.blockTimestamp || anchor.createdAt || ''),
            verificationUrl: (0, config_1.buildVerifyUrl)(hash),
            explorerUrl: txSig
                ? (0, config_1.buildExplorerUrl)(txSig, this.config.network)
                : '',
            status: this._mapStatus(String(anchor.status || 'pending')),
            name: anchor.metadata ? String(anchor.metadata) : undefined,
            contractAddress: this.config.contractAddress,
            network: this.config.network,
        };
    }
    _mapStatus(status) {
        const s = status.toLowerCase();
        if (s === 'confirmed')
            return 'confirmed';
        if (s === 'failed')
            return 'failed';
        if (s === 'revoked')
            return 'revoked';
        return 'pending';
    }
    _mapVerifyStatus(status) {
        const s = (status || '').toLowerCase();
        if (s === 'confirmed')
            return 'authentic';
        if (s === 'revoked')
            return 'revoked';
        if (s === 'not_found')
            return 'not_found';
        if (s === 'pending')
            return 'pending';
        return 'not_found';
    }
}
exports.SipHeron = SipHeron;
