// @sipheron/vdr-core/dist/client/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULTS = void 0;
exports.resolveConfig = resolveConfig;
exports.buildExplorerUrl = buildExplorerUrl;
exports.buildVerifyUrl = buildVerifyUrl;
exports.DEFAULTS = {
    network: 'devnet',
    timeout: 30000,
    retries: 3,
    baseUrls: {
        devnet: 'https://api.sipheron.com',
        mainnet: 'https://api.sipheron.com',
    },
    appUrl: 'https://app.sipheron.com',
    contractAddress: '6ecWPUK87zxwZP2pARJ75wbpCka92mYSGP1szrJxzAwo',
    explorerBase: {
        devnet: 'https://explorer.solana.com',
        mainnet: 'https://explorer.solana.com',
    },
};
function resolveConfig(config) {
    const network = config.network ?? exports.DEFAULTS.network;
    if (network === 'mainnet' && (!config.apiKey || typeof config.apiKey !== 'string')) {
        throw new Error('apiKey is required for mainnet. Get your free API key at https://app.sipheron.com');
    }
    return {
        apiKey: config.apiKey?.trim(),
        network,
        timeout: config.timeout ?? exports.DEFAULTS.timeout,
        retries: config.retries ?? exports.DEFAULTS.retries,
        baseUrl: config.baseUrl ?? exports.DEFAULTS.baseUrls[network],
        appUrl: exports.DEFAULTS.appUrl,
        contractAddress: exports.DEFAULTS.contractAddress,
        explorerBase: exports.DEFAULTS.explorerBase[network],
    };
}
function buildExplorerUrl(txSignature, network) {
    const base = exports.DEFAULTS.explorerBase[network];
    const cluster = network === 'devnet' ? '?cluster=devnet' : '';
    return `${base}/tx/${txSignature}${cluster}`;
}
function buildVerifyUrl(hash) {
    return `${exports.DEFAULTS.appUrl}/verify/${hash}`;
}
