// @sipheron/vdr-core/dist/client/http.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHttpClient = createHttpClient;
exports.withRetry = withRetry;
const axios_1 = __importDefault(require("axios"));
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const errors_1 = require("../errors");
function createHttpClient(config) {
    const instance = axios_1.default.create({
        baseURL: config.baseUrl,
        timeout: config.timeout,
        headers: {
            'Content-Type': 'application/json',
            'User-Agent': '@sipheron/vdr-core/0.1.0 Node/' + process.version,
        },
        httpsAgent: new https_1.default.Agent({ family: 4 }),
        httpAgent: new http_1.default.Agent({ family: 4 }),
    });
    // Response interceptor — map HTTP errors to typed SipHeron errors
    instance.interceptors.response.use((response) => response, (error) => {
        if (!error.response) {
            // Network error — no response received
            throw new errors_1.NetworkError(`Network request failed: ${error.message}`, error);
        }
        const { status, data } = error.response;
        const body = data;
        const message = String(body?.message || body?.error || error.message);
        switch (status) {
            case 401:
                throw new errors_1.AuthenticationError(message);
            case 400:
                throw new errors_1.ValidationError(message);
            case 404:
                throw new errors_1.AnchorNotFoundError(String(body?.hash || 'unknown'));
            case 429: {
                const errorCode = String(body?.error || '');
                if (errorCode === 'MONTHLY_QUOTA_EXCEEDED') {
                    const limit = body?.limit;
                    throw new errors_1.QuotaExceededError(Number(limit?.used ?? 0), Number(limit?.limit ?? 0), String(limit?.resetAt ?? ''), String(body?.upgrade ?? ''));
                }
                const retryAfter = Number(error.response.headers['retry-after'] ?? 60);
                throw new errors_1.RateLimitError(retryAfter, String(body?.plan ?? ''), String(body?.upgrade ?? ''));
            }
            case 500:
            case 502:
            case 503:
                throw new errors_1.NetworkError(`Server error (${status}): ${message}`);
            default:
                throw new errors_1.SipHeronError(message, 'API_ERROR', status);
        }
    });
    return instance;
}
/**
 * Retry a function with exponential backoff.
 */
async function withRetry(fn, maxAttempts, baseDelayMs = 1000) {
    let lastError;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
        try {
            return await fn();
        }
        catch (error) {
            lastError = error;
            // Do not retry on client errors (4xx) except rate limits
            if (error instanceof errors_1.SipHeronError) {
                const isRetryable = error instanceof errors_1.RateLimitError ||
                    error instanceof errors_1.NetworkError ||
                    (error.statusCode !== undefined && error.statusCode >= 500);
                if (!isRetryable)
                    throw error;
            }
            if (attempt === maxAttempts)
                break;
            const delay = baseDelayMs * Math.pow(2, attempt - 1);
            await sleep(delay);
        }
    }
    throw lastError;
}
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
