// @sipheron/vdr-core/dist/index.js
"use strict";
/**
 * @sipheron/vdr-core
 *
 * Core document verification library for SipHeron VDR.
 * Anchor any document's SHA-256 fingerprint permanently to Solana.
 * Verify authenticity anywhere, forever.
 *
 * @example
 * import { SipHeron } from '@sipheron/vdr-core'
 *
 * const sipheron = new SipHeron({ apiKey: 'your-key', network: 'devnet' })
 *
 * const anchor = await sipheron.anchor({ file, name: 'Contract v1' })
 * console.log(anchor.verificationUrl)
 *
 * const result = await sipheron.verify({ file })
 * console.log(result.authentic) // true
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionError = exports.SolanaConnectionError = exports.QuotaExceededError = exports.AnchorRevokedError = exports.ValidationError = exports.RateLimitError = exports.NetworkError = exports.HashMismatchError = exports.AnchorNotFoundError = exports.AuthenticationError = exports.SipHeronError = exports.anchorToSolana = exports.SOLANA_NETWORKS = exports.CONFIRMATION_DEPTH = exports.HASH_LENGTH = exports.MAX_BATCH_SIZE = exports.PROTOCOL_VERSION = exports.ANCHOR_SEED = exports.SIPHERON_PROGRAM_ID = exports.SIPHERON_CONTRACT = exports.estimateAnchorCost = exports.isValidTxSignature = exports.getExplorerUrl = exports.getCertificateProofText = exports.prepareCertificateData = exports.buildCertificateUrl = exports.generateCertificateId = exports.parseWebhookEvent = exports.verifyWebhookSignature = exports.deriveAnchorAddress = exports.verifyOnChain = exports.verifyLocally = exports.verifyHash = exports.normalizeHash = exports.isValidHash = exports.hashBase64 = exports.hashStream = exports.hashFile = exports.hashDocument = exports.SipHeron = void 0;
// ── Main client ──
var client_1 = require("./client");
Object.defineProperty(exports, "SipHeron", { enumerable: true, get: function () { return client_1.SipHeron; } });
// ── Standalone functions ──
var hash_1 = require("./hash");
Object.defineProperty(exports, "hashDocument", { enumerable: true, get: function () { return hash_1.hashDocument; } });
Object.defineProperty(exports, "hashFile", { enumerable: true, get: function () { return hash_1.hashFile; } });
Object.defineProperty(exports, "hashStream", { enumerable: true, get: function () { return hash_1.hashStream; } });
Object.defineProperty(exports, "hashBase64", { enumerable: true, get: function () { return hash_1.hashBase64; } });
Object.defineProperty(exports, "isValidHash", { enumerable: true, get: function () { return hash_1.isValidHash; } });
Object.defineProperty(exports, "normalizeHash", { enumerable: true, get: function () { return hash_1.normalizeHash; } });
var verify_1 = require("./verify");
Object.defineProperty(exports, "verifyHash", { enumerable: true, get: function () { return verify_1.verifyHashStandalone; } });
var verify_2 = require("./verify");
Object.defineProperty(exports, "verifyLocally", { enumerable: true, get: function () { return verify_2.verifyLocally; } });
Object.defineProperty(exports, "verifyOnChain", { enumerable: true, get: function () { return verify_2.verifyOnChain; } });
Object.defineProperty(exports, "deriveAnchorAddress", { enumerable: true, get: function () { return verify_2.deriveAnchorAddress; } });
var webhook_1 = require("./webhook");
Object.defineProperty(exports, "verifyWebhookSignature", { enumerable: true, get: function () { return webhook_1.verifyWebhookSignature; } });
Object.defineProperty(exports, "parseWebhookEvent", { enumerable: true, get: function () { return webhook_1.parseWebhookEvent; } });
// ── Certificate utilities ──
var certificate_1 = require("./certificate");
Object.defineProperty(exports, "generateCertificateId", { enumerable: true, get: function () { return certificate_1.generateCertificateId; } });
Object.defineProperty(exports, "buildCertificateUrl", { enumerable: true, get: function () { return certificate_1.buildCertificateUrl; } });
Object.defineProperty(exports, "prepareCertificateData", { enumerable: true, get: function () { return certificate_1.prepareCertificateData; } });
Object.defineProperty(exports, "getCertificateProofText", { enumerable: true, get: function () { return certificate_1.getCertificateProofText; } });
// ── Solana utilities ──
var anchor_1 = require("./anchor");
Object.defineProperty(exports, "getExplorerUrl", { enumerable: true, get: function () { return anchor_1.getExplorerUrl; } });
Object.defineProperty(exports, "isValidTxSignature", { enumerable: true, get: function () { return anchor_1.isValidTxSignature; } });
Object.defineProperty(exports, "estimateAnchorCost", { enumerable: true, get: function () { return anchor_1.estimateAnchorCost; } });
Object.defineProperty(exports, "SIPHERON_CONTRACT", { enumerable: true, get: function () { return anchor_1.SIPHERON_CONTRACT; } });
Object.defineProperty(exports, "SIPHERON_PROGRAM_ID", { enumerable: true, get: function () { return anchor_1.SIPHERON_PROGRAM_ID; } });
Object.defineProperty(exports, "ANCHOR_SEED", { enumerable: true, get: function () { return anchor_1.ANCHOR_SEED; } });
Object.defineProperty(exports, "PROTOCOL_VERSION", { enumerable: true, get: function () { return anchor_1.PROTOCOL_VERSION; } });
Object.defineProperty(exports, "MAX_BATCH_SIZE", { enumerable: true, get: function () { return anchor_1.MAX_BATCH_SIZE; } });
Object.defineProperty(exports, "HASH_LENGTH", { enumerable: true, get: function () { return anchor_1.HASH_LENGTH; } });
Object.defineProperty(exports, "CONFIRMATION_DEPTH", { enumerable: true, get: function () { return anchor_1.CONFIRMATION_DEPTH; } });
Object.defineProperty(exports, "SOLANA_NETWORKS", { enumerable: true, get: function () { return anchor_1.SOLANA_NETWORKS; } });
Object.defineProperty(exports, "anchorToSolana", { enumerable: true, get: function () { return anchor_1.anchorToSolana; } });
// ── Errors ──
var errors_1 = require("./errors");
Object.defineProperty(exports, "SipHeronError", { enumerable: true, get: function () { return errors_1.SipHeronError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return errors_1.AuthenticationError; } });
Object.defineProperty(exports, "AnchorNotFoundError", { enumerable: true, get: function () { return errors_1.AnchorNotFoundError; } });
Object.defineProperty(exports, "HashMismatchError", { enumerable: true, get: function () { return errors_1.HashMismatchError; } });
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return errors_1.NetworkError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return errors_1.RateLimitError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return errors_1.ValidationError; } });
Object.defineProperty(exports, "AnchorRevokedError", { enumerable: true, get: function () { return errors_1.AnchorRevokedError; } });
Object.defineProperty(exports, "QuotaExceededError", { enumerable: true, get: function () { return errors_1.QuotaExceededError; } });
Object.defineProperty(exports, "SolanaConnectionError", { enumerable: true, get: function () { return errors_1.SolanaConnectionError; } });
Object.defineProperty(exports, "TransactionError", { enumerable: true, get: function () { return errors_1.TransactionError; } });
