// @sipheron/vdr-core/dist/verify/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveAnchorAddress = exports.verifyOnChain = exports.verifyHashStandalone = exports.verifyLocally = exports.verifyBuffer = exports.compareHashes = void 0;
var compare_1 = require("./compare");
Object.defineProperty(exports, "compareHashes", { enumerable: true, get: function () { return compare_1.compareHashes; } });
Object.defineProperty(exports, "verifyBuffer", { enumerable: true, get: function () { return compare_1.verifyBuffer; } });
var verify_1 = require("./verify");
Object.defineProperty(exports, "verifyLocally", { enumerable: true, get: function () { return verify_1.verifyLocally; } });
Object.defineProperty(exports, "verifyHashStandalone", { enumerable: true, get: function () { return verify_1.verifyHashStandalone; } });
var onchain_1 = require("./onchain");
Object.defineProperty(exports, "verifyOnChain", { enumerable: true, get: function () { return onchain_1.verifyOnChain; } });
Object.defineProperty(exports, "deriveAnchorAddress", { enumerable: true, get: function () { return onchain_1.deriveAnchorAddress; } });
