// @sipheron/vdr-core/dist/verify/onchain.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveAnchorAddress = deriveAnchorAddress;
exports.verifyOnChain = verifyOnChain;
const web3_js_1 = require("@solana/web3.js");
const anchor = __importStar(require("@coral-xyz/anchor"));
const solana_1 = require("../anchor/solana");
const idl_json_1 = __importDefault(require("../anchor/idl.json"));
const hash_1 = require("../hash");
const errors_1 = require("../errors");
/**
 * Derive the Program Derived Address (PDA) for any anchor record.
 */
function deriveAnchorAddress(hashString, ownerPublicKey, network) {
    const hashBuffer = Buffer.from(hashString, 'hex');
    const programId = solana_1.SIPHERON_PROGRAM_ID[network];
    const [pda] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('hash_record'), hashBuffer, ownerPublicKey.toBuffer()], programId);
    return pda;
}
/**
 * Verify an anchor record directly by reading the Solana blockchain.
 * Does not require a SipHeron API key or server connection.
 */
async function verifyOnChain(options) {
    const { network, rpcUrl, ownerPublicKey } = options;
    if (!options.hash && !options.buffer) {
        throw new errors_1.ValidationError('Must provide either a hash or file buffer');
    }
    if (!ownerPublicKey) {
        throw new errors_1.ValidationError('ownerPublicKey is required for direct on-chain verification in v1 layout');
    }
    const hashString = options.buffer
        ? await (0, hash_1.hashDocument)(options.buffer)
        : options.hash.toLowerCase();
    const connectionUrl = rpcUrl || solana_1.SOLANA_NETWORKS[network];
    const connection = new web3_js_1.Connection(connectionUrl, 'confirmed');
    const ownerPk = typeof ownerPublicKey === 'string'
        ? new web3_js_1.PublicKey(ownerPublicKey)
        : ownerPublicKey;
    const pda = deriveAnchorAddress(hashString, ownerPk, network);
    const programId = solana_1.SIPHERON_PROGRAM_ID[network];
    const provider = new anchor.AnchorProvider(connection, new anchor.Wallet(web3_js_2.Keypair.generate()), // Dummy wallet for read-only
    {});
    const program = new anchor.Program(idl_json_1.default, provider);
    try {
        const record = await program.account.hashRecord.fetch(pda);
        return {
            authentic: !record.isRevoked,
            hash: hashString,
            pda: pda.toBase58(),
            owner: record.owner.toBase58(),
            timestamp: record.timestamp.toNumber(),
            isRevoked: record.isRevoked,
            metadata: record.metadata,
        };
    }
    catch (err) {
        if (err.message && err.message.includes('Account does not exist')) {
            return {
                authentic: false,
                hash: hashString,
            };
        }
        throw new errors_1.SolanaConnectionError(`Failed to verify on chain: ${err.message}`);
    }
}
// Dummy import for Keypair inside read-only wallet above
const web3_js_2 = require("@solana/web3.js");
