// @sipheron/vdr-core/dist/verify/verify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyLocally = verifyLocally;
exports.verifyHashStandalone = verifyHashStandalone;
const hash_1 = require("../hash");
const compare_1 = require("./compare");
const errors_1 = require("../errors");
/**
 * Perform a standalone local verification.
 * Compares a document or hash against a known anchor hash.
 * No network call — purely local comparison.
 *
 * Use this when you have the anchor hash and want to verify
 * a document locally without making an API call.
 *
 * @param file - Document as Buffer, OR
 * @param hash - Pre-computed SHA-256 hash
 * @param anchorHash - The hash stored in the anchor record
 * @returns Local verification result
 */
async function verifyLocally(fileOrHash, anchorHash) {
    if (!anchorHash || !(0, hash_1.isValidHash)((0, hash_1.normalizeHash)(anchorHash))) {
        throw new errors_1.ValidationError('anchorHash must be a valid 64-character SHA-256 hex string');
    }
    let computedHash;
    if (Buffer.isBuffer(fileOrHash)) {
        computedHash = await (0, hash_1.hashDocument)(fileOrHash);
    }
    else {
        const normalized = (0, hash_1.normalizeHash)(fileOrHash);
        if (!(0, hash_1.isValidHash)(normalized)) {
            throw new errors_1.ValidationError('Invalid hash format. Must be 64 hex characters.');
        }
        computedHash = normalized;
    }
    return {
        authentic: (0, compare_1.compareHashes)(computedHash, anchorHash),
        computedHash,
        anchorHash: (0, hash_1.normalizeHash)(anchorHash),
    };
}
/**
 * Standalone verifyHash function.
 * Calls the SipHeron public verification API without authentication.
 * Use this for simple hash lookups without the full client.
 *
 * @param hash - 64-character SHA-256 hash
 * @param apiBaseUrl - API base URL. Default: https://api.sipheron.com
 */
async function verifyHashStandalone(hash, apiBaseUrl = 'https://api.sipheron.com') {
    const normalized = (0, hash_1.normalizeHash)(hash);
    if (!(0, hash_1.isValidHash)(normalized)) {
        throw new errors_1.ValidationError('Invalid hash format. Must be a 64-character SHA-256 hex string.');
    }
    const response = await fetch(`${apiBaseUrl}/api/verify`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ hash: normalized }),
    });
    const data = (await response.json());
    return {
        authentic: data.authentic === true,
        status: data.status === 'CONFIRMED' ? 'authentic'
            : data.status === 'REVOKED' ? 'revoked'
                : data.error === 'NOT_FOUND' ? 'not_found'
                    : 'not_found',
        hash: normalized,
        verifiedAt: data.verified_at || new Date().toISOString(),
        anchor: data.anchor,
    };
}
