// @sipheron/vdr-core/dist/anchor/anchor.js
"use strict";
/**
 * Core anchoring logic exposed as standalone functions.
 * These functions use the SipHeron HTTP API to anchor documents.
 *
 * For most use cases, use the SipHeron client class instead:
 *   const sipheron = new SipHeron({ apiKey })
 *   await sipheron.anchor({ file })
 *
 * These standalone functions are for advanced use cases where
 * you need more control over the anchoring process.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashDocument = void 0;
exports.prepareAnchor = prepareAnchor;
exports.mapToAnchorResult = mapToAnchorResult;
const hash_1 = require("../hash");
Object.defineProperty(exports, "hashDocument", { enumerable: true, get: function () { return hash_1.hashDocument; } });
const errors_1 = require("../errors");
const config_1 = require("../client/config");
/**
 * Prepare anchor data from AnchorOptions.
 * Computes hash if file is provided, validates hash if string is provided.
 * Returns the normalized hash ready for API submission.
 */
async function prepareAnchor(options) {
    if (!options.file && !options.hash) {
        throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string).');
    }
    let hash;
    if (options.file) {
        hash = await (0, hash_1.hashDocument)(options.file);
    }
    else {
        hash = (0, hash_1.normalizeHash)(options.hash);
        if (!(0, hash_1.isValidHash)(hash)) {
            throw new errors_1.ValidationError('Invalid hash format. Must be 64 hex characters (SHA-256).');
        }
    }
    return {
        hash,
        metadata: options.name || null,
    };
}
/**
 * Map raw API response to AnchorResult.
 */
function mapToAnchorResult(data, network = 'devnet') {
    const record = (data.record || data);
    const hash = String(record.hash || '');
    const txSig = String(record.txSignature || '');
    return {
        id: String(record.id || ''),
        hash,
        transactionSignature: txSig,
        blockNumber: Number(record.blockNumber || 0),
        timestamp: String(record.blockTimestamp || record.createdAt || ''),
        verificationUrl: (0, config_1.buildVerifyUrl)(hash),
        explorerUrl: txSig ? (0, config_1.buildExplorerUrl)(txSig, network) : '',
        status: ['confirmed', 'failed', 'revoked', 'pending']
            .find(s => s === record.status?.toString().toLowerCase()) || 'pending',
        name: record.metadata ? String(record.metadata) : undefined,
        contractAddress: '6ecWPUK87zxwZP2pARJ75wbpCka92mYSGP1szrJxzAwo',
        network,
    };
}
