// @sipheron/vdr-core/dist/anchor/solana.js
"use strict";
/**
 * Solana-specific utilities for the vdr-core library.
 * Includes explicit program constants.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EXPLORER_URLS = exports.SIPHERON_CONTRACT = exports.CONFIRMATION_DEPTH = exports.HASH_LENGTH = exports.MAX_BATCH_SIZE = exports.PROTOCOL_VERSION = exports.ANCHOR_SEED = exports.SIPHERON_PROGRAM_ID = exports.SIPHERON_CONTRACT_BASE58 = exports.SOLANA_NETWORKS = void 0;
exports.getExplorerUrl = getExplorerUrl;
exports.isValidTxSignature = isValidTxSignature;
exports.estimateAnchorCost = estimateAnchorCost;
const web3_js_1 = require("@solana/web3.js");
/**
 * Note: This module provides utilities for working with Solana
 * transaction data — it does NOT handle transaction signing
 * or wallet key management. Those operations happen server-side
 * in the SipHeron platform.
 *
 * For direct Solana integration, use the SipHeron API.
 */
exports.SOLANA_NETWORKS = {
    devnet: 'https://api.devnet.solana.com',
    mainnet: 'https://api.mainnet-beta.solana.com',
};
exports.SIPHERON_CONTRACT_BASE58 = '6ecWPUK87zxwZP2pARJ75wbpCka92mYSGP1szrJxzAwo';
exports.SIPHERON_PROGRAM_ID = {
    devnet: new web3_js_1.PublicKey(exports.SIPHERON_CONTRACT_BASE58),
    mainnet: new web3_js_1.PublicKey(exports.SIPHERON_CONTRACT_BASE58)
};
exports.ANCHOR_SEED = 'anchor';
exports.PROTOCOL_VERSION = 1;
exports.MAX_BATCH_SIZE = 500;
exports.HASH_LENGTH = 64;
exports.CONFIRMATION_DEPTH = 32;
exports.SIPHERON_CONTRACT = exports.SIPHERON_CONTRACT_BASE58;
exports.EXPLORER_URLS = {
    devnet: {
        tx: (sig) => `https://explorer.solana.com/tx/${sig}?cluster=devnet`,
        address: (addr) => `https://explorer.solana.com/address/${addr}?cluster=devnet`,
        block: (slot) => `https://explorer.solana.com/block/${slot}?cluster=devnet`,
    },
    mainnet: {
        tx: (sig) => `https://explorer.solana.com/tx/${sig}`,
        address: (addr) => `https://explorer.solana.com/address/${addr}`,
        block: (slot) => `https://explorer.solana.com/block/${slot}`,
    },
};
/**
 * Build a Solana Explorer URL for a transaction.
 */
function getExplorerUrl(txSignature, network = 'devnet') {
    return exports.EXPLORER_URLS[network].tx(txSignature);
}
/**
 * Validate a Solana transaction signature format.
 * Signatures are base58-encoded 64-byte values — typically 87-88 chars.
 */
function isValidTxSignature(sig) {
    return typeof sig === 'string' &&
        sig.length >= 43 &&
        sig.length <= 128 &&
        /^[1-9A-HJ-NP-Za-km-z]+$/.test(sig);
}
/**
 * Estimate the approximate SOL cost of anchoring a document.
 * Based on average Solana transaction fees as of 2026.
 */
function estimateAnchorCost() {
    const lamports = 5000; // ~0.000005 SOL per transaction
    return {
        lamports,
        sol: lamports / 1000000000,
        usdApprox: '< $0.01',
    };
}
