// @sipheron/vdr-core/dist/errors/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionError = exports.SolanaConnectionError = exports.QuotaExceededError = exports.AnchorRevokedError = exports.ValidationError = exports.RateLimitError = exports.NetworkError = exports.HashMismatchError = exports.AnchorNotFoundError = exports.AuthenticationError = exports.SipHeronError = void 0;
/**
 * Base error class for all SipHeron errors.
 * All errors have a machine-readable code for programmatic handling.
 */
class SipHeronError extends Error {
    constructor(message, code, statusCode) {
        super(message);
        this.code = code;
        this.statusCode = statusCode;
        this.name = 'SipHeronError';
        // Maintains proper stack trace in V8
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.SipHeronError = SipHeronError;
/**
 * Thrown when the API key is missing, invalid, or revoked.
 */
class AuthenticationError extends SipHeronError {
    constructor(message = 'Invalid or missing API key.') {
        super(message, 'AUTHENTICATION_ERROR', 401);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * Thrown when no anchor record is found for the given hash.
 */
class AnchorNotFoundError extends SipHeronError {
    constructor(hash) {
        super(`No anchor record found for hash: ${hash.slice(0, 16)}...`, 'ANCHOR_NOT_FOUND', 404);
        this.hash = hash;
        this.name = 'AnchorNotFoundError';
    }
}
exports.AnchorNotFoundError = AnchorNotFoundError;
/**
 * Thrown when the document hash does not match the anchored hash.
 * This means the document has been modified since anchoring.
 */
class HashMismatchError extends SipHeronError {
    constructor(computedHash, anchoredHash) {
        super('Document hash does not match anchor record. The document may have been modified.', 'HASH_MISMATCH');
        this.computedHash = computedHash;
        this.anchoredHash = anchoredHash;
        this.name = 'HashMismatchError';
    }
}
exports.HashMismatchError = HashMismatchError;
/**
 * Thrown when a network request fails.
 */
class NetworkError extends SipHeronError {
    constructor(message, originalError) {
        super(message, 'NETWORK_ERROR');
        this.originalError = originalError;
        this.name = 'NetworkError';
    }
}
exports.NetworkError = NetworkError;
/**
 * Thrown when the API rate limit is exceeded.
 * Check retryAfter to know when to retry.
 */
class RateLimitError extends SipHeronError {
    constructor(retryAfter, plan, upgradeUrl = 'https://app.sipheron.com/dashboard/billing') {
        super(`Rate limit exceeded. Retry after ${retryAfter} seconds.`, 'RATE_LIMIT_ERROR', 429);
        this.retryAfter = retryAfter;
        this.plan = plan;
        this.upgradeUrl = upgradeUrl;
        this.name = 'RateLimitError';
    }
}
exports.RateLimitError = RateLimitError;
/**
 * Thrown when request parameters are invalid.
 */
class ValidationError extends SipHeronError {
    constructor(message) {
        super(message, 'VALIDATION_ERROR', 400);
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
/**
 * Thrown when an anchor has been revoked.
 */
class AnchorRevokedError extends SipHeronError {
    constructor(hash, revokedAt, reason) {
        super(`Anchor for hash ${hash.slice(0, 16)}... has been revoked.`, 'ANCHOR_REVOKED');
        this.hash = hash;
        this.revokedAt = revokedAt;
        this.reason = reason;
        this.name = 'AnchorRevokedError';
    }
}
exports.AnchorRevokedError = AnchorRevokedError;
/**
 * Thrown when the monthly quota is exceeded.
 */
class QuotaExceededError extends SipHeronError {
    constructor(used, limit, resetAt, upgradeUrl = 'https://app.sipheron.com/dashboard/billing') {
        super(`Monthly anchor quota exceeded (${used}/${limit}). Resets at ${resetAt}.`, 'QUOTA_EXCEEDED', 429);
        this.used = used;
        this.limit = limit;
        this.resetAt = resetAt;
        this.upgradeUrl = upgradeUrl;
        this.name = 'QuotaExceededError';
    }
}
exports.QuotaExceededError = QuotaExceededError;
/**
 * Thrown when connecting to a Solana RPC node fails.
 */
class SolanaConnectionError extends SipHeronError {
    constructor(message) {
        super(message, 'SOLANA_CONNECTION_ERROR', 503);
        this.name = 'SolanaConnectionError';
    }
}
exports.SolanaConnectionError = SolanaConnectionError;
/**
 * Thrown when a direct on-chain Solana transaction fails or is rejected.
 */
class TransactionError extends SipHeronError {
    constructor(message) {
        super(message, 'TRANSACTION_ERROR', 400);
        this.name = 'TransactionError';
    }
}
exports.TransactionError = TransactionError;
